### PR TITLE
Transition to the new upgrade framework for upgrade tests

### LINKED ISSUE #628: Make the e2e test viable on different platforms
**Problem**
Currently, `ko` is bound to the `linux/amd64` platform during the e2e test. There seems no way of getting it configured on the fly.

**[Persona:](https://github.com/knative/eventing/blob/main/docs/personas.md)**
Which persona is this feature for?
People who want to run the e2e test on a different platform

**Exit Criteria**
To open up using an environment variable like below:

```
diff --git a/test/e2e-common.sh b/test/e2e-common.sh
index fc7544d6..f46adcbb 100755
--- a/test/e2e-common.sh
+++ b/test/e2e-common.sh
@@ -35,6 +35,7 @@ export SYSTEM_NAMESPACE=${TEST_NAMESPACE}
 # We will use only one namespace, when Knative supports both components can coexist under one namespace.
 export TEST_EVENTING_NAMESPACE="knative-eventing"
 export TEST_RESOURCE="knative"
+export KO_FLAGS="${KO_FLAGS:-}"
 
 # Boolean used to indicate whether to generate serving YAML based on the latest code in the branch KNATIVE_SERVING_REPO_BRANCH.
 GENERATE_SERVING_YAML=0
@@ -158,7 +159,7 @@ function install_operator() {
   download_latest_release
   header "Installing Knative operator"
   # Deploy the operator
-  ko apply -f config/
+  ko apply ${KO_FLAGS} -f config/
   wait_until_pods_running default || fail_test "Operator did not come up"
 }

```

**Time Estimate (optional):**
How many developer-days do you think this may take to resolve?

**Additional context (optional)**
Add any other context about the feature request here.

### PR DESCRIPTION
**Pull request by houshengbo**
Fixes: #428

## Proposed Changes

* This PR rewrites the upgrade tests under the new framework.
* This PR only fixes the upgrade prow for operator, not for serving and eventing.
* This PR does not contain any golang code refactoring for the existing code, except copying and pasting the existing code.
* This PR stops running any tests in the prow pull-knative-operator-eventing-upgrade-tests and pull-knative-operator-serving-upgrade-tests. We will fix them in future PRs.

### CODE REVIEW

File: test/e2e-common.sh [hunk 1]
```diff
@@ -31,14 +31,15 @@ readonly ISTIO_VERSION="stable"
 readonly ISTIO_MESH=0
 # Namespaces used for tests
 export TEST_NAMESPACE="${TEST_NAMESPACE:-knative-operator-testing}"
-export SYSTEM_NAMESPACE=${TEST_NAMESPACE}           # knative-serving
-export TEST_EVENTING_NAMESPACE=${TEST_NAMESPACE}    # knative-eventing
-export TEST_RESOURCE="knative"    # knative-eventing
+export SYSTEM_NAMESPACE=${TEST_NAMESPACE}
+# TODO: remove when components can coexist in same namespace
```
Comment (Cynocracy) [inline test/e2e-common.sh:35]: Is there an immediate plan for this? I don't know if we need a TODO in place unless it's planned soon.

File: test/e2e-eventing-upgrade-tests.sh [hunk 2]
```diff
@@ -105,8 +105,8 @@ header "Running tests for Knative Operator"
 # Operator tests here will make sure that all the Knative deployments reach the desired states and operator CR is
 # in ready state.
 cd ${OPERATOR_DIR}
-go_test_e2e -tags=postupgrade -timeout=${TIMEOUT} ./test/upgrade \
-  --preservingversion="${PREVIOUS_SERVING_RELEASE_VERSION}" --preeventingversion="${PREVIOUS_EVENTING_RELEASE_VERSION}" || failed=1
+# go_test_e2e -tags=postupgrade -timeout=${TIMEOUT} ./test/upgrade \
```
Comment (Cynocracy) [inline test/e2e-eventing-upgrade-tests.sh:108]: uncomment?

File: test/e2e-common.sh (hunk 1 above)
Comment (houshengbo) [reply test/e2e-common.sh:35]: I do not think so. serving and eventing have to be under different namespaces.

File: test/e2e-eventing-upgrade-tests.sh (hunk 2 above)
Comment (houshengbo) [reply test/e2e-eventing-upgrade-tests.sh:108]: Yes. I will fix the serving and eventing prows in others PRs. Just uncomment to make them pass.

File: test/upgrade/installation/shell.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+func LatestRelease() pkgupgrade.Operation {
+	return install("ServingEventingLatestRelease", "create_latest_custom_resource")
+}
+
+// PreviousRelease installs Knative Serving and Eventing from the previous stable release.
+func PreviousRelease() pkgupgrade.Operation {
+	return install("ServingEventingPreviousRelease", "install_previous_knative")
+}
+
+func install(installName, shellFunc string) pkgupgrade.Operation {
```
Comment (evankanderson) [inline test/upgrade/installation/shell.go:39]: I'm slightly terrified at how deep this goes between `go` and `bash`, but I don't have a well-considered alternative right now.

**Review by evankanderson — COMMENTED** [review]
Two bash comments; I wish I had a good suggestion on how to simplify the tests.

File: test/upgrade/installation/shell.go (hunk 3 above)
Comment (houshengbo) [reply test/upgrade/installation/shell.go:39]: Golang code calls the function in bash scripts.
There is sh file to import.
Based on knative's need, it calls the function to upgrade and downgrade, and the tests, after running them.

File: test/e2e-upgrade-tests.sh [hunk 4]
```diff
@@ -33,135 +33,18 @@
 
 export GO111MODULE=auto
 
-source $(dirname $0)/e2e-common.sh
+source "$(dirname "${BASH_SOURCE[0]}")/e2e-common.sh"
 
-# TODO: remove when components can coexist in same namespace
```
[the remark this answers is not in the document]
Comment (houshengbo) [reply test/e2e-upgrade-tests.sh:38]: This is not something that will happen soon IMO.
I just removed it based on @Cynocracy comment.

File: test/upgrade/installation/shell.go (hunk 3 above)
Comment (Cynocracy) [reply test/upgrade/installation/shell.go:39]: The scary bit is going Bash -> Golang -> Bash in layers, makes it hard to reason about the test, iiuc.

File: test/upgrade/preupgrade.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+			t.Fatalf("Failed to get KnativeEventing CR: %v", err)
+		}
+		resources.SetKodataDir()
+		// Based on the status.version, get the deployment resources.
+		defer os.Unsetenv(common.KoEnvKey)
+		manifest, err := common.InstalledManifest(keventing)
+		if err != nil {
+			t.Fatalf("Failed to get the manifest for Knative: %v", err)
+		}
+		expectedDeployments := resources.GetExpectedDeployments(manifest)
```
Comment (Cynocracy) [inline test/upgrade/preupgrade.go:115]: I have a hard time understanding these assertions, maybe a comment would be helpful?

**Review by Cynocracy — COMMENTED** [review]
/lgtm

**Pull request merged**